### PR TITLE
fix: robust layer-count parsing for nested checkpoint prefixes (complements #336)

### LINKED ISSUE #335: ValueError: invalid literal for int() with base 10: 'layers'
Bug: Number of layers parsed incorrectly when layer names not defined.

Root cause: the model layers key from the index file after split by "." will have the following components: "model" [0], "layers" [1], "<N layer number>" [2].
In some cases, like Qwen/Qwen3.6-27, the elements after split: "ge_model" [0], "layers" [1], "<N layer number>" [2], since the code does an interesting heuristic to cut a prefix.

When layer_names parameter is missing, it uses the correct index after split by dot - 2, but when this parameter is defined, it uses 1.

I could not locate the index file spec in safetensors docs: https://huggingface.co/docs/safetensors/en/index

I checked the following models:
https://huggingface.co/Qwen/Qwen3.6-27B/blob/main/model.safetensors.index.json
https://huggingface.co/timdettmers/guanaco-33b-merged/blob/main/pytorch_model.bin.index.json
https://huggingface.co/Qwen/Qwen3-235B-A22B/blob/main/model.safetensors.index.json
https://huggingface.co/moonshotai/Kimi-K3/blob/main/model.safetensors.index.json (different key arrangement)
https://huggingface.co/zai-org/GLM-5.2/blob/main/model.safetensors.index.json

My understanding that there must be a model index file that actually will result in the index 1 to correctly point to a layer number, but I could not find such.

**Question:** Overall, the parsing logic and the format are loosely defined, which leads to brittle parsing logic. Do you know if there is a case where it would require to take the element at 1 position after the key split by dot or we can simply fix it by changing the split index to 2?

<details>
<summary>Reproduction code</summary>

```python
from airllm import AutoModel

MAX_LENGTH = 128
# just pass a hugging face repo id — works with almost any popular model:
model = AutoModel.from_pretrained("Qwen/Qwen3.6-27B")

# go bigger with the exact same one line:
#model = AutoModel.from_pretrained("Qwen/Qwen3-235B-A22B")     # 235B, runs in ~3GB
#model = AutoModel.from_pretrained("deepseek-ai/DeepSeek-V3")  # 671B, runs in ~12GB

# or use a model's local path...
# model = AutoModel.from_pretrained("/home/ubuntu/.cache/huggingface/hub/models--Qwen--Qwen3-32B/snapshots/...")

input_text = [
    'What is the capital of United States?',
    #'I like',
]

input_tokens = model.tokenizer(input_text,
                               return_tensors="pt",
                               return_attention_mask=False,
                               truncation=True,
                               max_length=MAX_LENGTH,
                               padding=False)

generation_output = model.generate(
    input_tokens['input_ids'].cuda(),
    max_new_tokens=20,
    use_cache=True,
    return_dict_in_generate=True)

output = model.tokenizer.decode(generation_output.sequences[0])

print(output)
```

</details>

<details>
<summary>Traceback</summary>

```python
using generic AirLLM streaming model for architecture: Qwen3_5ForConditionalGeneration
Fetching 14 files: 100%|██████████| 14/14 [00:00<00:00, 319.98it/s]
Traceback (most recent call last):
File "/home/pc/test/model.py", line 5, in <module>
model = AutoModel.from_pretrained("Qwen/Qwen3.6-27B")
File "/home/pc/test/.venv/lib/python3.14/site-packages/airllm/auto_model.py", line 61, in from_pretrained
return class_(pretrained_model_name_or_path, *inputs, **kwargs)
File "/home/pc/test/.venv/lib/python3.14/site-packages/airllm/airllm_base.py", line 131, in __init__
self.model_local_path, self.checkpoint_path = find_or_create_local_splitted_path(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
    model_local_path_or_repo_id,
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                              ...<3 lines>...
hf_token=hf_token,
^^^^^^^^^^^^^^^^^^
delete_original=delete_original)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pc/test/.venv/lib/python3.14/site-packages/airllm/utils.py", line 554, in find_or_create_local_splitted_path
return Path(hf_cache_path), split_and_save_layers(hf_cache_path, layer_shards_saving_path,
                                                  ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                  compression=compression, layer_names=layer_names,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
delete_original=delete_original, repo_id=model_local_path_or_repo_id, hf_token=hf_token)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pc/test/.venv/lib/python3.14/site-packages/airllm/utils.py", line 284, in split_and_save_layers
n_layers = len(set([int(k[len(layer_names['layer_prefix']):].split('.')[1]) for k in index.keys() if layer_names['layer_prefix'] in k]))
~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'layers'
```

</details>

<details>
<summary>Debug screenshots</summary>
Image

Image

</details>

### PR DESCRIPTION
**Pull request by HEETMEHTA18**
Fixes #335 (both branches -- see comment on #336 for why that PR's
fix is scoped to only one branch).

`utils.py`'s layer-counting logic assumed every key matching
`model.layers` starts with that prefix, so it always read the layer
index from a fixed split position. Checkpoints where the LM sits
under a wrapper module (e.g. `language_model.model.layers.0...`,
common in VLM/multimodal architectures) still match the substring
filter but break the positional assumption, causing:

    ValueError: invalid literal for int() with base 10: 'layers'

This applies the same `.find()` + `.isdigit()` guard already used
for per-expert MoE key parsing in `airllm_base.py`, to both the
`layer_names is None` and custom-`layer_names` branches, so it's
correct for arbitrary nesting depth rather than one fixed position.

Verified against both a normal flat checkpoint and a nested
wrapper-style checkpoint (see test snippet in code comments / commit
message).

## Why not just the #336 approach

#336 moves `split('.')[1]` to `split('.')[2]` in the custom-`layer_names`
branch only. Two gaps:

1. The `layer_names is None` branch (default architecture path) is not
   touched and crashes on the same nested-key shape reproduced in #335.
2. Even in the `layer_names` branch, a fixed split index is the same
   class of bug at a different depth. For Key K3's
   `language_model.model.layers.0.self_attn...` keys the post-prefix
   segment is `.0.self_attn...`, where `[1]` is the layer index but `[2]`
   is `'self_attn'` — i.e. the `[2]` shift would *also* raise the exact
   bug on the K3 layout exercised by `air_llm/tests/test_kimi_k3_split.py`.

The `.find()` + `.isdigit()` approach handles both branches at any
nesting depth. `air_llm/tests/test_kimi_k3_split.py` passes unchanged
with this change.

### CODE REVIEW

File: air_llm/airllm/utils.py [hunk 1]
```diff
@@ -278,10 +278,29 @@ def split_and_save_layers(checkpoint_path, layer_shards_saving_path=None, splitt
             f"No model weights found under {checkpoint_path}. Expected one of: "
             f"model.safetensors(.index.json) or pytorch_model.bin(.index.json).")
 
+    def _count_layers(index_keys, marker):
```
Comment (Tradunsky) [inline air_llm/airllm/utils.py]: It might make sense to create various tests using different models prefix keys to ensure they are all covered.

**Comment by HEETMEHTA18** [comment]
@Tradunsky  Thanks for the suggestion! I’ll add tests covering different model/checkpoint prefix patterns to make sure the layer-count parsing handles various nesting depths and model architectures correctly. I’ll make the changes and push the updated tests to the same branch.

File: air_llm/airllm/utils.py (hunk 1 above)
Comment (HEETMEHTA18) [reply air_llm/airllm/utils.py]: Done — added a dedicated test file (`air_llm/tests/test_layer_count.py`) covering a matrix of the key prefixes seen across the supported architectures plus the nested shapes that triggered the bug:

- `model.layers.N...` (generic / Llama / Baichuan / InternLM / Mistral / Mixtral / Qwen2)
- `transformer.h.N...` (Qwen1)
- `transformer.encoder.layers.N...` (ChatGLM)
- `language_model.model.layers.N...` (Kimi K3)
- Nested VLM shape `language_model.model.layers.N...` and deeper multi-level nesting, plus the `ge_model.layers.N...` shape from #335
- Edge cases: non-layer keys, empty index, non-numeric segment, duplicate indices

To make it directly unit-testable I also extracted `_count_layers` to a module-level private helper. All 13 new tests pass, and the existing `test_kimi_k3_split` suite (8 tests) still passes unchanged.

File: air_llm/airllm/utils.py [hunk 2]
```diff
@@ -232,6 +232,20 @@ def link_or_copy_file(src, dst):
...
+def _count_layers(index_keys, marker):
+    # Layer index = the digit segment right after `marker` in each key. .find() + .isdigit()
+    # (mirroring the MoE expert parser in airllm_base.py) makes this correct for VLM-style
+    # keys like "language_model.model.layers.0.self_attn.q_proj.weight", where the index is
+    # not at a fixed split position -- the previous code crashed with
+    # "ValueError: invalid literal for int() with base 10: 'layers'".
+    nums = set()
+    for k in index_keys:
+        pos = k.find(marker)
+        if pos != -1 and (head := k[pos + len(marker):].split('.', 1)[0]).isdigit():
```
Comment (Tradunsky) [inline air_llm/airllm/utils.py]: Although this version is shorter, it has higher cognitive load to understand and debug.

I would suggest keeping the previous version.

File: air_llm/tests/test_layer_count.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    def test_chatglm_nested_layer_prefix(self):
+        # ChatGLM uses transformer.encoder.layers.N... (airllm_chatglm.py)
+        prefix = 'transformer.encoder.layers'
+        keys = layer_keys(prefix, 6, 'self_attention.dense.weight') + ['transformer.word_embeddings.weight']
+        self.assertEqual(_count_layers(keys, prefix + '.'), 6)
+
+    def test_kimi_k3_language_model_prefix(self):
+        # Kimi K3 nests the decoder under language_model (airllm_kimi_k3.py)
+        prefix = 'language_model.model.layers'
+        keys = [f'{prefix}.{i}.block_sparse_moe.experts.0.w1.weight' for i in range(3)]
```
Comment (Tradunsky) [inline air_llm/tests/test_layer_count.py]:
```suggestion
        keys = layer_keys(prefix, 3, 'block_sparse_moe.experts.0.w1.weight')
```

**Review by Tradunsky — COMMENTED** [review]
A few suggestions. Overall looks good! Thank you for adding tests!

File: air_llm/airllm/utils.py (hunk 2 above)
Comment (HEETMEHTA18) [reply air_llm/airllm/utils.py]: Agreed — readability wins over a couple of saved lines here. Reverted to the explicit loop version in 1e15a5a. Thanks for the review.

File: air_llm/tests/test_layer_count.py (hunk 3 above)
Comment (HEETMEHTA18) [reply air_llm/tests/test_layer_count.py]: Applied — now uses layer_keys(prefix, 3, 'block_sparse_moe.experts.0.w1.weight') in 1e15a5a. All 21 tests still pass.

**Comment by HEETMEHTA18** [comment]
Hey @lyogavin @Tradunsky — this PR is ready to merge. It fixes #335 by replacing the fixed-position layer-index parsing in `utils.py` with the same `.find()` + `.isdigit()` guard already used for MoE expert keys in `airllm_base.py`, covering both branches at any nesting depth. Includes a matrix of tests across all supported architecture prefixes, and the existing `test_kimi_k3_split` suite passes unchanged. All checks green, no conflicts — happy to address any further feedback.

**Review by Tradunsky — APPROVED** [review]
Thank you for considering my suggestions.

Ready to fly! 🚢

**Comment by HEETMEHTA18** [comment]
@Tradunsky  thank you!!  for providing more suggestions to the changes that i have committed!!
